### PR TITLE
Add log2 to Work

### LINKED ISSUE #1326: Add `log2` method to `Work` and `Target` returning `f64`
Core displays some values as log2 and it could be useful to have this conversion too. Returning `f64` delegates to Rust formatting so we don't have to worry about it. I guess newtypes are not important here but it should be documented it's mainly intended for displaying.

See https://github.com/rust-bitcoin/rust-bitcoin/pull/1197#discussion_r998171557_

### PR DESCRIPTION
**Pull request by jirijakes**
Adds method `Work::log2()` providing value equivalent to Bitcoin Core's `log2_work` in its logs. Fixes #1326.

Questions:

- The original issue (#1326) also suggests to add log2 to Target but it does not seem to be meaningful, does it? Bitcoin Core, to which the issue refers, also displays only log2 of work.
- Although work should not be 0, the type allows it. In this case, log2 would return -inf. I think we could leave it like that but if there are suggestions to deal with it in a different way, please let me know.

### CODE REVIEW

File: bitcoin/src/pow.rs [hunk 1]
```diff
@@ -110,6 +110,16 @@ impl Work {
 
     /// Converts this [`Work`] to [`Target`].
     pub fn to_target(self) -> Target { Target(self.0.inverse()) }
+
+    /// Returns log2 of this work. The result inherently suffers from a loss of precision
```
Comment (Kixunil) [inline bitcoin/src/pow.rs:114]: Needs new paragraph after the dot.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Just doc issue, the rest looks good.

**Comment by Kixunil** [comment]
> but it does not seem to be meaningful, does it?

Oh, right. It wouldn't hurt but also probably not useful.

> 0, the type allows it. In this case, log2 would return -inf.

Yeah, -inf makes sense. I think last time we concluded preventing zero doesn't make sense.

File: bitcoin/src/pow.rs (hunk 1 above)
Comment (jirijakes) [reply bitcoin/src/pow.rs]: Done, thanks.

File: bitcoin/src/pow.rs [hunk 2]
```diff
@@ -110,6 +110,16 @@ impl Work {
...
+    /// and is, therefore, meant to be used mainly for informative and displaying purposes,
+    /// similarly to Bitcoin Core's `log2_work` output in its logs.
+    pub fn log2(self) -> f64 {
+        let U256(high, low) = self.0;
+        // 2^128 * high + low
+        let double = (3402823669209385e23_f64 * high as f64) + (low as f64);
```
Comment (sanket1729) [inline bitcoin/src/pow.rs:122]: verified this value is correct.

**Review by sanket1729 — COMMENTED** [review]
LTGM, can you please squash the commits?

File: bitcoin/src/pow.rs (hunk 2 above)
Comment (sanket1729) [reply bitcoin/src/pow.rs:122]:
```
>>> float(2**128)
3.402823669209385e+38
>>> 3402823669209385/10**15
3.402823669209385
>>> 
```

**Comment by jirijakes** [comment]
> can you please squash the commits?

Yes, squashed, @sanket1729.

**Review by sanket1729 — APPROVED** [review]
ACK 812701317ce94df210684d71280e1bfc33f6fa53

**Comment by apoelstra** [comment]
Hah, cool! I was pretty apprehensive about this feature, but just manually coercing 2**128 to a f64 is a pretty clever solution.

**Comment by apoelstra** [comment]
Frustratingly CI is failing because `log2` is not available on ARM...I think we should feature-gate this method, but I'm not sure how exactly.

**Review by apoelstra — DISMISSED** [review]
ACK 812701317ce94df210684d71280e1bfc33f6fa53

**Comment by jirijakes** [comment]
> Frustratingly CI is failing because `log2` is not available on ARM.

I am not familiar at all with other architectures. How about `log10` or `ln`, might they be available?

**Comment by jirijakes** [comment]
> How about `log10`, might it be available?

I managed to run the embedded test and seems like the answer is negative.

**Comment by apoelstra** [comment]
Heh, maybe just `log`? But I'm doubtful.

**Comment by Kixunil** [comment]
This has nothing to do with ARM, logarithm just isn't in `core`, it needs `std`.

I suggest we just `#[cfg(feature = "std")] #[cfg_attr(docsrs, doc(cfg(feature = "std"))]` it. If someone wants to submit software-implemented log2 I'll be happy to ACK.

**Comment by jirijakes** [comment]
> This has nothing to do with ARM, logarithm just isn't in `core`, it needs `std`.

Ah, makes sense.

> I suggest we just `#[cfg(feature = "std")] #[cfg_attr(docsrs, doc(cfg(feature = "std"))]` it. 

Done.

**Comment by Kixunil** [comment]
Will ACK when you squash the commits. :)

**Comment by jirijakes** [comment]
> Will ACK when you squash the commits. :)

And done!

**Review by Kixunil — APPROVED** [review]
ACK df90c5024282526dc7316afea220a545f9f9fa95

**Review by tcharding — APPROVED** [review]
ACK df90c5024282526dc7316afea220a545f9f9fa95

**Comment by tcharding** [comment]
Nice one man, I started to investigate writing a "software implementation" of this the other week and couldn't do it :)

**Review by apoelstra — APPROVED** [review]
ACK df90c5024282526dc7316afea220a545f9f9fa95

**Pull request merged**